src/linear_algebra/finsupp.lean

```diff
@@ -141,247 +97,320 @@ begin
-
-open linear_map submodule
-
-instance : vector_space γ (α →₀ β) :=
-{ to_module := finsupp.module }
-
-lemma is_basis_single {f : α → set β} (hf : ∀a, is_basis γ (f a)) :
-  is_basis γ (⋃a, single a '' f a) :=
-⟨linear_independent_single $ assume a, (hf a).1,
-  by simp only [span_Union, span_single_image, (hf _).2, map_top, supr_lsingle_range]⟩
+lemma mem_supported {s : set α} (p : α →₀ β) : p ∈ (supported β γ s) ↔ ↑p.support ⊆ s :=
+iff.rfl
 
-end vector_space
+lemma mem_supported' {s : set α}  (p : α →₀ β) :
+  p ∈ supported β γ s ↔ ∀ x ∉ s, p x = 0 :=
+by haveI := classical.dec_pred (λ (x : α), x ∈ s);
+   simp [mem_supported, set.subset_def, not_imp_comm]
 
-section dim
-universes u v
-variables {α : Type u} {β : Type u} {γ : Type v}
-variables [decidable_eq α] [decidable_eq β] [discrete_field γ] [add_comm_group β] [vector_space γ β]
+lemma single_mem_supported {s : set α} {a : α} (b : β) (h : a ∈ s) :
+  single a b ∈ supported β γ s :=
+set.subset.trans support_single_subset (set.singleton_subset_iff.2 h)
 
-lemma dim_eq : vector_space.dim γ (α →₀ β) = cardinal.mk α * vector_space.dim γ β :=
+lemma supported_eq_span_single [decidable_eq γ] [has_one β] (s : set α) :
+  supported γ γ s = span γ ((λ i, single i 1) '' s) :=
 begin
-  rcases exists_is_basis γ β with ⟨bs, hbs⟩,
-  rw [← hbs.mk_eq_dim, ← (is_basis_single (λa:α, hbs)).mk_eq_dim, cardinal.mk_Union_eq_sum_mk],
-  { simp only [cardinal.mk_image_eq (injective_single.{u u} _), cardinal.sum_const] },
-  { refine assume i j h, disjoint_image_image (assume b hb c hc, _),
-    simp only [(≠), single_eq_single_iff, not_or_distrib, not_and_distrib],
-    have : (0:β) ∉ bs := zero_not_mem_of_linear_independent (zero_ne_one : (0:γ) ≠ 1) hbs.1,
-    exact ⟨or.inl h, or.inl (assume eq, this $ eq ▸ hb)⟩ }
+  refine (span_eq_of_le _ _ (le_def'.2 $ λ l hl, _)).symm,
+  { rintro _ ⟨_, hp, rfl ⟩ , exact single_mem_supported γ 1 hp },
+  { rw ← l.sum_single,
+    refine sum_mem _ (λ i il, _),
+    convert @smul_mem γ (α →₀ γ) _ _ _ _ (single i 1) (l i) _,
+    { simp },
+    apply subset_span,
+    apply set.mem_image_of_mem _ (hl il) }
 end
 
-end dim
+variables (β γ)
 
-end finsupp
+def restrict_dom (s : set α) [decidable_pred (λ x, x ∈ s)]: (α →₀ β) →ₗ supported β γ s :=
+linear_map.cod_restrict _
+  { to_fun := filter (∈ s),
+    add := λ l₁ l₂, filter_add,
+    smul := λ a l, filter_smul }
+  (λ l, (mem_supported' _ _).2 $ λ x, filter_apply_neg (∈ s) l)
+
+variables {β γ}
 
-namespace lc
-universes u v w
-variables (α : Type v) {β : Type u} {γ : Type w}
-variables [ring α]
-variables [add_comm_group β] [module α β]
-variables [add_comm_group γ] [module α γ]
+set_option class.instance_max_depth 50
+@[simp] theorem restrict_dom_apply (s : set α) [decidable_pred (λ x, x ∈ s)] (l : α →₀ β) :
+  ((restrict_dom β γ s : α →₀ β →ₗ supported β γ s) l : α →₀ β) = finsupp.filter (∈ s) l := rfl
+set_option class.instance_max_depth 32
```

Comment:
You can put the set_option in a section to have it revert to the default after the section

Reply:
Done.